File: dart/test/onboarding_test.dart
import 'package:gha_indie_worker_lib_core/gha_indie_worker_lib_core.dart';
import 'package:test/test.dart';

List<S> _reachable<S>(S start, List<S> Function(S) next) {
  final seen = <S>[start];
  for (var i = 0; i < seen.length; i++) {
    for (final state in next(seen[i])) {
      if (!seen.contains(state)) seen.add(state);
    }
  }
  return seen;
}

void main() {
  test('org edges are deterministic', () {
    for (final edge in orgTransitions) {
      final targets = orgTransitions
          .where((e) => e.$1 == edge.$1 && e.$2 == edge.$2)
          .toList();
      expect(targets.length, 1, reason: '${edge.$1} + ${edge.$2}');
    }
  });

  test('no self loops in either machine', () {
    for (final edge in orgTransitions) {
      expect(edge.$1, isNot(edge.$3));
    }
    for (final edge in userTransitions) {
      expect(edge.$1, isNot(edge.$3));
    }
  });

  test('every state is reachable from created', () {
    expect(
      _reachable(OrgState.created, orgNextStates).toSet(),
      OrgState.values.toSet(),
    );
    expect(
      _reachable(UserState.created, userNextStates).toSet(),
      UserState.values.toSet(),
    );
  });

  test('the happy path walks created to active', () {
    var state = OrgState.created;
    const path = [
      OnboardingEvent.profileCompleted,
      OnboardingEvent.billingLinked,
      OnboardingEvent.githubAppInstalled,
      OnboardingEvent.firstWorkflowPlanned,
      OnboardingEvent.activated,
    ];
    for (final event in path) {
      final outcome = advanceOrg(state, event);
      expect(outcome, isA<Advanced<OrgState>>(), reason: '$state + $event');
      state = (outcome as Advanced<OrgState>).state;
    }
    expect(state, OrgState.active);
    expect(orgMayRunWorkflows(state), isTrue);
  });

  test('replaying an event is idempotent, not an error', () {
    expect(
      advanceOrg(OrgState.orgProfile, OnboardingEvent.profileCompleted),
      isA<AlreadyThere<OrgState>>(),
    );
  });

  test('an event from the other machine is unknown', () {
    final outcome = advanceOrg(OrgState.created, OnboardingEvent.emailVerified);
    expect((outcome as Rejected<OrgState>).reason, Rejection.unknownEvent);
  });

  test('wire values match the contract authorities', () {
    expect(OrgState.githubAppInstalled.wire, 'github-app-installed');
    expect(UserState.profileComplete.wire, 'profile-complete');
    expect(OrgState.fromWire('billing-linked'), OrgState.billingLinked);
    expect(OrgState.fromWire('nope'), isNull);
  });
}

File: typescript/test/onboarding.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';

import {
  ORG_STATES, USER_STATES, ORG_EVENTS, USER_EVENTS,
  ORG_TRANSITIONS, USER_TRANSITIONS,
  advanceOrg, advanceUser, orgNextStates, userNextStates, orgMayRunWorkflows,
} from '../src/onboarding.mjs';

const reachable = (start, next) => {
  const seen = [start];
  for (let i = 0; i < seen.length; i += 1) {
    for (const state of next(seen[i])) if (!seen.includes(state)) seen.push(state);
  }
  return seen;
};

test('org edges are deterministic', () => {
  for (const [from, event] of ORG_TRANSITIONS) {
    const targets = ORG_TRANSITIONS.filter(([f, e]) => f === from && e === event);
    assert.equal(targets.length, 1, `${from} + ${event}`);
  }
});

test('user edges are deterministic', () => {
  for (const [from, event] of USER_TRANSITIONS) {
    const targets = USER_TRANSITIONS.filter(([f, e]) => f === from && e === event);
    assert.equal(targets.length, 1, `${from} + ${event}`);
  }
});

test('no self loops', () => {
  for (const [from, , to] of [...ORG_TRANSITIONS, ...USER_TRANSITIONS]) assert.notEqual(from, to);
});

test('every state is reachable from created', () => {
  assert.deepEqual(reachable('created', orgNextStates).sort(), [...ORG_STATES].sort());
  assert.deepEqual(reachable('created', userNextStates).sort(), [...USER_STATES].sort());
});

test('every edge names a known state and event', () => {
  for (const [from, event, to] of ORG_TRANSITIONS) {
    assert.ok(ORG_STATES.includes(from) && ORG_STATES.includes(to), `${from}->${to}`);
    assert.ok(ORG_EVENTS.includes(event), event);
  }
  for (const [from, event, to] of USER_TRANSITIONS) {
    assert.ok(USER_STATES.includes(from) && USER_STATES.includes(to), `${from}->${to}`);
    assert.ok(USER_EVENTS.includes(event), event);
  }
});

test('the happy path walks created to active', () => {
  let state = 'created';
  for (const event of ['profile-completed', 'billing-linked', 'github-app-installed', 'first-workflow-planned', 'activated']) {
    const result = advanceOrg(state, event);
    assert.equal(result.outcome, 'advanced', `${state} + ${event}`);
    state = result.state;
  }
  assert.equal(state, 'active');
  assert.ok(orgMayRunWorkflows(state));
});

test('replaying an event is idempotent, not an error', () => {
  assert.deepEqual(advanceOrg('org-profile', 'profile-completed'), { outcome: 'already-there', state: 'org-profile' });
});

test('skipping a step is rejected', () => {
  assert.equal(advanceOrg('created', 'activated').reason, 'not-allowed-here');
});

test('an event from the other machine is unknown', () => {
  assert.equal(advanceOrg('created', 'email-verified').reason, 'unknown-event');
  assert.equal(advanceUser('created', 'billing-linked').reason, 'unknown-event');
});

test('suspension is reachable from every org state except itself', () => {
  for (const state of ORG_STATES) {
    if (state === 'suspended') continue;
    assert.equal(advanceOrg(state, 'suspended').outcome, 'advanced', state);
  }
});

test('every state/event pair is decided and never throws', () => {
  for (const state of ORG_STATES) {
    for (const event of ORG_EVENTS) {
      assert.ok(['advanced', 'already-there', 'rejected'].includes(advanceOrg(state, event).outcome));
    }
  }
  for (const state of USER_STATES) {
    for (const event of USER_EVENTS) {
      assert.ok(['advanced', 'already-there', 'rejected'].includes(advanceUser(state, event).outcome));
    }
  }
});
